File: javascript/scripts/check-routes.mjs
#!/usr/bin/env node
import fs from "node:fs";
import path from "node:path";
import { fileURLToPath } from "node:url";

const packageRoot = path.resolve(path.dirname(fileURLToPath(import.meta.url)), "..");
const manifest = JSON.parse(fs.readFileSync(path.join(packageRoot, "scripts", "routes.json"), "utf8"));

function readTree(dir) {
  const out = [];
  for (const entry of fs.readdirSync(dir, { withFileTypes: true })) {
    const full = path.join(dir, entry.name);
    if (entry.isDirectory()) {
      out.push(...readTree(full));
    } else {
      out.push(full);
    }
  }
  return out;
}

const sources = Object.fromEntries(
  readTree(path.join(packageRoot, "src"))
    .filter((file) => file.endsWith(".ts"))
    .map((file) => [file, fs.readFileSync(file, "utf8")]),
);

function pathNeedles(operationPath) {
  const staticPrefix = operationPath.split("{")[0].replace(/\/$/, "");
  return [operationPath.replace(/\{[^}]+\}/g, ""), staticPrefix].filter(Boolean);
}

function hasPathCoverage(operationPath, methodHint) {
  const needles = pathNeedles(operationPath);
  return Object.values(sources).some((content) => {
    if (!needles.some((needle) => content.includes(needle))) {
      return false;
    }
    if (methodHint === "POST" && operationPath.includes("/cancel")) {
      return content.includes("cancel");
    }
    if (methodHint === "GET" && operationPath.includes("/events")) {
      return content.includes("events");
    }
    if (methodHint === "GET" && operationPath.includes("/children")) {
      return content.includes("children");
    }
    if (methodHint === "GET" && operationPath.includes("/archive")) {
      return content.includes("/archive");
    }
    if (methodHint === "GET" && operationPath.endsWith("/volume")) {
      return content.includes("/volume");
    }
    if (methodHint === "GET" && operationPath.includes("/grep")) {
      return content.includes("/grep");
    }
    if (methodHint === "POST" && operationPath.includes("/summarize")) {
      return content.includes("/summarize");
    }
    if (operationPath.includes("/file_lines/")) {
      return content.includes("/file_lines/");
    }
    if (methodHint === "DELETE" && operationPath.includes("/paths/")) {
      return content.includes("/paths/");
    }
    return true;
  });
}

const failures = [];
for (const operation of manifest.operations) {
  if (!hasPathCoverage(operation.path, operation.method)) {
    failures.push(`Missing coverage for ${operation.method} ${operation.path} (${operation.symbol})`);
  }
}

if (failures.length > 0) {
  console.error("@agent-api/sdk route coverage failed:\n" + failures.map((line) => `- ${line}`).join("\n"));
  process.exit(1);
}

console.log(`@agent-api/sdk route coverage OK (${manifest.operations.length} operations)`);

File: python/scripts/check_routes.py
#!/usr/bin/env python3
from __future__ import annotations

import json
import pathlib
import sys

ROOT = pathlib.Path(__file__).resolve().parents[1]
manifest = json.loads((ROOT / "scripts" / "routes.json").read_text())
sources = {path: path.read_text() for path in (ROOT / "src" / "agent_api").rglob("*.py")}


def path_needles(operation_path: str) -> list[str]:
    import re

    static_prefix = operation_path.split("{", 1)[0].rstrip("/")
    needles = [re.sub(r"\{[^}]+\}", "", operation_path), static_prefix]
    return [needle for needle in needles if needle]


def has_path_coverage(operation_path: str, method: str) -> bool:
    needles = path_needles(operation_path)
    for content in sources.values():
        if not any(needle in content for needle in needles):
            continue
        if method == "POST" and "/cancel" in operation_path and "cancel" not in content:
            continue
        if method == "GET" and "/events" in operation_path and "events" not in content:
            continue
        if method == "GET" and "/children" in operation_path and "children" not in content:
            continue
        if method == "GET" and "/archive" in operation_path and "/archive" not in content:
            continue
        if method == "GET" and operation_path.endswith("/volume") and "/volume" not in content:
            continue
        if method == "GET" and "/grep" in operation_path and "/grep" not in content:
            continue
        if "POST" in method and "/summarize" in operation_path and "/summarize" not in content:
            continue
        if method == "GET" and "/file_lines/" in operation_path and "/file_lines/" not in content:
            continue
        if method == "PATCH" and "/file_lines/" in operation_path and "/file_lines/" not in content:
            continue
        if method == "DELETE" and "/paths/" in operation_path and "/paths/" not in content:
            continue
        return True
    return False


failures = [
    f'Missing coverage for {op["method"]} {op["path"]} ({op["symbol"]})'
    for op in manifest["operations"]
    if not has_path_coverage(op["path"], op["method"])
]

if failures:
    print("cloudsway-agent route coverage failed:", file=sys.stderr)
    for line in failures:
        print(f"- {line}", file=sys.stderr)
    sys.exit(1)

print(f"cloudsway-agent route coverage OK ({len(manifest['operations'])} operations)")
